torch/csrc/jit/constants.h

```diff
@@ -35,12 +35,37 @@ TORCH_API at::optional<IValue> toIValue(Value* v);
 
 // if a value is a constant then try to turn into type T using the
 // same rules as the interpreter
+// use DEFINE_CONSTANT_AS to define a new case
 template<typename T>
-at::optional<T> constant_as(Value* v) {
-  if(auto ivalue = toIValue(v)) {
-    return ivalue->to<T>();
-  }
-  return at::nullopt;
+at::optional<T> constant_as(Value* v) = delete;
+
+// without isName() exception will be thrown if the IValue is the wrong type
+#define DEFINE_CONSTANT_AS(type, name)                                      \
+template<>                                                                  \
+inline at::optional<type> constant_as(Value* v) {                           \
+  if (at::optional<IValue> ivalue = toIValue(v)) {                          \
+    if (ivalue->is ## name()) {                                             \
+      return ivalue->to<type>();                                            \
+    }                                                                       \
+  }                                                                         \
+  return at::nullopt;                                                       \
 }
 
+DEFINE_CONSTANT_AS(at::Tensor, Tensor)
```

Comment:
Let's not do this. It changes the behavior everywhere else constant_as is used and adds a bunch of code to replace a very simple function:
```
if(auto r = toIValue(v)) {
  if(r.isInt())
      return r.toInt();
  return at::nullopt;
}
```

Reply:
In getIntListAttribute constant_as is called on the input with Shared<jit::IntList>, and then invoked with <int64_t>. In the case that it is the input is an in64_t this will throw an exception. It looks like that function is not being called anywhere currently, but it seems to be an easy pattern to fall into.

Reply:
getIntListAttribute is dead, let's just delete it. We should keep the convention that the optional return indicates whether the value was a constant, and not whether it was a constant of the type you wanted. We shouldn't encourage developers using polymorphic constants as it will cause people to miss cases where a constant can have a different type.